lib/src/commit.rs

```diff
@@ -155,6 +155,10 @@ impl Commit {
 
     /// A slow (but cached) way to get the full verification.
     pub fn verification(&self) -> SignResult<Option<Verification>> {
+        if !self.store().signer().show_signatures {
+            return None.transpose();
+        }
```

Comment:
Personally I think that having the templater think that there's no signature without the `--show-signature` flag (or the config) is convenient as we can have the pretty checkmarks in the default config and they just wont show up without the flag.

However, we could do a quick check for the signature presence (the above `is_signed` which is just a null check, as fast as it gets) and return `Verification::Unknown` here so that even without `--show-signature` the templater draws the question marks showing that commits have unchecked signatures on them.

Or even have a distinct Verification variant specifically for that case, so that the templater can draw something less ugly than question marks (an asterisk, perhaps) - this idea actually has grown on me a lot, to show that there *is* something in the signature header, we just wont run the heavy verifications without opt-in

---
Alternatively, we can go the inverse route and when show-signatures is off avoid reading the signature from git entirely, straight in the core where I've already implemented the signature reading without thinking about performance of it too much - I think this is what git does

^ @martinvonz @yuja pinging you on this random review in the middle of ~~nowhere~~ the PR as I've got too deep into designing the bit I haven't given much thought previously

Reply:
> Personally I think that having the templater think that there's no signature without the --show-signature flag (or the config) is convenient as we can have the pretty checkmarks in the default config and they just wont show up without the flag.

Do you mean something like `template-aliases.embed_signature = true|false` which can be toggled by user?

I don't know the goal of this option nor what would be the best way around, but I think it's weird that `Signer` had `show_signature: bool` flag.

Reply:
@necauqua

> here so that even without `--show-signature` the templater draws the question marks showing that commits have unchecked signatures on them.

Personally I don't want the log to render anything at all about signatures unless I explicitly ask for them. It adds noise and I don't think it gives any valuable information.

> Alternatively, we can go the inverse route and when show-signatures is off avoid reading the signature from git entirely, straight in the core where I've already implemented the signature reading without thinking about performance of it too much - I think this is what git does

This makes the most sense to me, but I don't know how to go about implementing any of this. Is this something we could do in a follow up PR?

I also think it's weird for the signer to have the `show_signature` flag but I'm not very familiar with the code-base and it isn't obvious to me how to have this information flow through all the way to where it is currently used.

I suspect to achieve that it would take some significant changes that I am probably not the best person to make.

Reply:
> Personally I don't want the log to render anything at all about signatures unless I explicitly ask for them.

If it's just to silence the default output, can we split the default template change `cli/src/config/templates.toml` to another PR?

This PR is getting bigger, and it wouldn't be ideal to run review & fix loop at this size.

Reply:
> @yuja
> Do you mean something like `template-aliases.embed_signature = true|false` [...]

The code where I left the comment is the one that makes the signer just return no verification if `show_signatures` config/flag is false - with my proposal being returning `Verification::Unknown`/`Verification::SpecialVariantAboutOpaqueSignatureBeingPresent` so that the templater knows there is a signature header

> I don't want the log to render anything at all about signatures

I mean you can change the template however you want, we're talking more about the default one here

> It adds noise and I don't think it gives any valuable information

It's not that much noise (~~one~~ three characters in the log for the status) and it _is_ valuable?.
Like you'd see that some commit is signed somehow (for free, performance-wise) and you might want to run the `--show-signature` log, where without the character giving it away you wouldn't know

But yeah, this can be done and/or discussed later in another PR, same with the sig header reading optimization (I don't think it impacts performance too much or it'd have been noticed, at least on repos without tens of thousands of _signed_ commits)

Reply:
> The code where I left the comment is the one that makes the signer just return no verification if `show_signatures` config/flag is false - with my proposal being returning `Verification::Unknown`/`Verification::SpecialVariantAboutOpaqueSignatureBeingPresent` so that the templater knows there is a signature header

Got it. In order to bypass slow verification process, doing that makes sense. If the verification is optional, it might be even better to make `Commit` object return `secure_sig`, and let the caller to do the verification as needed. The result could be cached by the templater (if it is the solo consumer.)

> It adds noise and I don't think it gives any valuable information

For people who aren't interested in cryptographic signature, a few characters and emojis can be visual noise. Anyway, the purpose of `--show-signature` is to suppress the template output, I think it should be dealt with the templating machinery.

Reply:
> If the verification is optional, it might be even better to make Commit object return secure_sig, and let the caller to do the verification as needed.

Technically I agree with this but I have no idea how to get a handle on the `UserSettings` struct from the context of the templater. I would need this to check the `show-settings` property.

I'm also very new to both rust and this project - can I get some pointers on how to achieve this?

> The result could be cached by the templater

The result is actually cached by the signer AFAIK

Reply:
> Technically I agree with this but I have no idea how to get a handle on the `UserSettings` struct from the context of the templater.

I don't think the templater needs to know whether signature should be verified. It's rather controlled by the template content. But let's discuss this later maybe in separate PR.

> > The result could be cached by the templater
> 
> The result is actually cached by the signer AFAIK (so it shouldn't matter where this is called it will always be cached.)

I mean the repo-level cache could be moved to the templater if the templater is solo consumer.